dkg/pedersen/handler.go

```diff
@@ -530,3 +607,299 @@ func (h *Handler) announceDkgPublicKey(out mino.Sender, from mino.Address) error
+	dela.Logger.Trace().Msgf("%v is handling deals from other nodes", h.me)
+
+	addrsNew := resharingRequest.GetAddrsNew()
+	var addrsOld []mino.Address
+
+	numReceivedDeals := 0
+	expectedDeals := 0
+	// common nodes receive one deal less than the new nodes and they have the old addresses saved in their state
+	if commonNode {
+		addrsOld = h.startRes.GetParticipants()
+		expectedDeals = len(addrsOld) - 1
+
+	} else {
+		addrsOld = resharingRequest.GetAddrsOld()
+		expectedDeals = len(addrsOld)
+	}
+
+	// we find the union of the old and new address sets to aviod from sending a message to the common nodes multiple times
+	addrsAll := unionOfTwoSlices(addrsNew, addrsOld)
+
+	for numReceivedDeals < expectedDeals {
+		select {
+
+		case df := <-h.dealsResharing:
+			// the new nodes that are not part of the old committee should create a dkg after receiving their first deal
+			if !commonNode && numReceivedDeals == 0 {
+				c := &pedersen.Config{
+					Suite:        suite,
+					Longterm:     h.privKey,
+					OldNodes:     resharingRequest.GetPubkeysOld(),
+					NewNodes:     resharingRequest.GetPubkeysNew(),
+					PublicCoeffs: df.dealResharing.GetPublicCoeffs(),
+					Threshold:    resharingRequest.GetTNew(),
+					OldThreshold: resharingRequest.GetTOld(),
+				}
+
+				newDkg, err := pedersen.NewDistKeyHandler(c)
+				if err != nil {
+					return xerrors.Errorf("failed to create the new dkg for %s: %v",
+						h.me, err)
+				}
+				h.dkg = newDkg
+			}
+
+			deal := df.dealResharing.GetDeal()
+			err := h.handleDeal(&deal, df.from, addrsAll, out)
+			if err != nil {
+				dela.Logger.Warn().Msgf("%s failed to handle received deal from %s: %v",
+					h.me, from, err)
+				return xerrors.Errorf("failed to handle received deal: %v", err)
+			}
+
+			dela.Logger.Trace().Msgf("%s handled deal from %s",
+				h.me, df.from)
+			numReceivedDeals++
+
+		case <-ctx.Done():
+			dela.Logger.Error().Msgf("%s timed out while receiving deals", h.me)
+			return xerrors.Errorf("timed out while receiving deals")
+		}
+	}
+
+	dela.Logger.Debug().Msgf("%v received all the expected deals", h.me)
+
+	return nil
+}
+
+///////// auxiliary functions
+// gets an address and a slice of addresses and returns true if that address is in the slice.
+//  this function is called for cheking whether an old committee member is in the new commmitte as well or not
+func addrIsInSlice(addr mino.Address, addrSlice []mino.Address) bool {
+	for _, i := range addrSlice {
+		if addr.Equal(i) {
+			return true
+		}
+	}
+	return false
+}
+
+// gets the list of the old committee members and new committee members and returns the union
+func unionOfTwoSlices(addrsSlice1 []mino.Address, addrsSlice2 []mino.Address) []mino.Address {
+	addrsAll := addrsSlice1
+	for _, i := range addrsSlice2 {
+		exist := false
+		for _, j := range addrsSlice1 {
+			if i == j {
+				exist = true
+				break
+			}
+		}
+		if !exist {
+			addrsAll = append(addrsAll, i)
+		}
+	}
```

Comment:
this has complexity O(n*m) if `n` and `m` are the lengths of the slices.
You can do this in O(n+m) and have simpler code , see https://siongui.github.io/2018/03/10/go-set-of-all-elements-in-two-arrays/

Reply:
You are trading space for time, but at least this is more readable. Also the example should use `make(map[int]struct{})`, and not `make(map[int]bool)`.

Reply:
so should I use the example or not?